[Pre4] Blazor WASM performance profiling

guardrex opened the pull request:
Fixes #35230
Addresses #34437

## Notes

* This PR is refactoring the *Performance* articles into a node (folder) and splitting the large `performance.md` file into a set of shorter articles. The only one that you need to review is the `profiling.md` article.
* I pulled the `ds-router` remarks from the draft content.
* I'm not on nightly builds. The untested examples are taken from your issue/PR remarks. I can go nightly if needed for testing, but I think you'll be able to suggest corrections based on what's on the PR.
* If there are many changes, you might want to edit the `profiling.md` markdown file directly.
* Do you want to include guidance on how the profiling configuration can be set up to only run in `Debug` configuration (i.e., applying one or more `Condition` attributes and/or targets+properties) so that profiling won't accidentally be left enabled for a published (released) app?


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

<details><summary><strong>Toggle expand/collapse</strong></summary><br/>

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/components/event-handling.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/components/event-handling.md) | [aspnetcore/blazor/components/event-handling](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/event-handling?branch=pr-en-us-35243) |
| [aspnetcore/blazor/components/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/components/index.md) | [aspnetcore/blazor/components/index](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/index?branch=pr-en-us-35243) |
| [aspnetcore/blazor/components/lifecycle.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/components/lifecycle.md) | [aspnetcore/blazor/components/lifecycle](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?branch=pr-en-us-35243) |
| [aspnetcore/blazor/components/rendering.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/components/rendering.md) | [aspnetcore/blazor/components/rendering](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/rendering?branch=pr-en-us-35243) |
| [aspnetcore/blazor/components/templated-components.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/components/templated-components.md) | [aspnetcore/blazor/components/templated-components](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/components/templated-components?branch=pr-en-us-35243) |
| [aspnetcore/blazor/globalization-localization.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/globalization-localization.md) | [aspnetcore/blazor/globalization-localization](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/globalization-localization?branch=pr-en-us-35243) |
| [aspnetcore/blazor/host-and-deploy/configure-linker.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/host-and-deploy/configure-linker.md) | [aspnetcore/blazor/host-and-deploy/configure-linker](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/configure-linker?branch=pr-en-us-35243) |
| [aspnetcore/blazor/host-and-deploy/configure-trimmer.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/host-and-deploy/configure-trimmer.md) | [aspnetcore/blazor/host-and-deploy/configure-trimmer](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/configure-trimmer?branch=pr-en-us-35243) |
| [aspnetcore/blazor/javascript-interoperability/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/javascript-interoperability/index.md) | [aspnetcore/blazor/javascript-interoperability/index](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/index?branch=pr-en-us-35243) |
| [aspnetcore/blazor/performance/app-download-size.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/performance/app-download-size.md) | [aspnetcore/blazor/performance/app-download-size](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/performance/app-download-size?branch=pr-en-us-35243) |
| [aspnetcore/blazor/performance/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/performance/index.md) | [aspnetcore/blazor/performance/index](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/performance/index?branch=pr-en-us-35243) |
| [aspnetcore/blazor/performance/javascript-interoperability.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/performance/javascript-interoperability.md) | [aspnetcore/blazor/performance/javascript-interoperability](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/performance/javascript-interoperability?branch=pr-en-us-35243) |
| [aspnetcore/blazor/performance/rendering.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/performance/rendering.md) | [aspnetcore/blazor/performance/rendering](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/performance/rendering?branch=pr-en-us-35243) |
| [aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md) | [aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/performance/webassembly-browser-developer-tools-diagnostics?branch=pr-en-us-35243) |
| [aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md) | [aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/performance/webassembly-event-pipe-diagnostics?branch=pr-en-us-35243) |
| [aspnetcore/blazor/webassembly-lazy-load-assemblies.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/blazor/webassembly-lazy-load-assemblies.md) | [aspnetcore/blazor/webassembly-lazy-load-assemblies](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/webassembly-lazy-load-assemblies?branch=pr-en-us-35243) |
| [aspnetcore/diagnostics/bl0001.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/diagnostics/bl0001.md) | [aspnetcore/diagnostics/bl0001](https://review.learn.microsoft.com/en-us/aspnet/core/diagnostics/bl0001?branch=pr-en-us-35243) |
| [aspnetcore/diagnostics/bl0004.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/diagnostics/bl0004.md) | [aspnetcore/diagnostics/bl0004](https://review.learn.microsoft.com/en-us/aspnet/core/diagnostics/bl0004?branch=pr-en-us-35243) |
| [aspnetcore/performance/diagnostic-tools.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/performance/diagnostic-tools.md) | [aspnetcore/performance/diagnostic-tools](https://review.learn.microsoft.com/en-us/aspnet/core/performance/diagnostic-tools?branch=pr-en-us-35243) |
| [aspnetcore/performance/memory.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/performance/memory.md) | [aspnetcore/performance/memory](https://review.learn.microsoft.com/en-us/aspnet/core/performance/memory?branch=pr-en-us-35243) |
| [aspnetcore/performance/overview.md](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/performance/overview.md) | [aspnetcore/performance/overview](https://review.learn.microsoft.com/en-us/aspnet/core/performance/overview?branch=pr-en-us-35243) |
| [aspnetcore/toc.yml](https://github.com/dotnet/AspNetCore.Docs/blob/7a3322d868bd001719f997b085bef4f1c0cab277/aspnetcore/toc.yml) | [aspnetcore/toc](https://review.learn.microsoft.com/en-us/aspnet/core/toc?branch=pr-en-us-35243) |

</details>


<!-- PREVIEW-TABLE-END -->

pavelsavara commented on aspnetcore/blazor/performance/app-download-size.md:
```diff
@@ -0,0 +1,108 @@
+
+Blazor WebAssembly's runtime includes the following .NET features that can be disabled for a smaller payload size.
+
+Blazor WebAssembly carries globalization resources required to display values, such as dates and currency, in the user's culture. If the app doesn't require localization, you may [configure the app to support the invariant culture](xref:blazor/globalization-localization#invariant-globalization), which is based on the `en-US` culture.
```
Perhaps you can mention `<InvariantGlobalization>true</InvariantGlobalization>` directly in the text here.

pavelsavara commented on aspnetcore/blazor/performance/app-download-size.md:
````diff
@@ -0,0 +1,108 @@
+<BlazorWebAssemblyPreserveCollationData>false</BlazorWebAssemblyPreserveCollationData>
+</PropertyGroup>
+```
+
````
Please add link to https://github.com/dotnet/runtime/blob/main/src/mono/wasm/features.md for more flags

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
```diff
@@ -0,0 +1,226 @@
+
+* [Ahead-of-time (AOT) compilation](xref:blazor/tooling/webassembly#ahead-of-time-aot-compilation)
+* Code interpolation
+* [JIT (Just-In-Time) interpolation](https://developer.mozilla.org/docs/Glossary/Just_In_Time_Compilation) 
```
We don't do any JIT for WASM, instead we do `jiterp` but it's too low level detail. Please drop the JIT mention.

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
````diff
@@ -0,0 +1,226 @@
+
+```xml
+<PropertyGroup>
+  <WasmProfilers>browser;</WasmProfilers>
````
You can use `<WasmProfilers>browser;</WasmProfilers>` without AOT. 

Also you can dev-tools profiler with AOT and without `<WasmProfilers>browser;</WasmProfilers>`.
To be able to see AOT method names in the browser tools, you need to install [DWARF chrome extension](https://chromewebstore.google.com/detail/cc++-devtools-support-dwa/pdcpmagijalfljmkmjngeonclgbbannb)

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
````diff
@@ -0,0 +1,226 @@
+`<WasmBuildNative>` | `false` | `true` | Controls building the native executable.
+
+```xml
+<PropertyGroup>
````
Enabling profilers has negative size and performance impact. You don't want to publish such build into production.

One possible option is to enable the whole section via condition.

`<PropertyGroup Condition="'$(MyAppWithEnabledProfiler)' == 'true'">`

And then pass `/p:MyAppWithEnabledProfiler=true` when building from command-line.

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
````diff
@@ -0,0 +1,226 @@
+</script>
+```
+
+## Log profiling for memory troubleshooting
````
Please drop the whole log profiler section. It's not supported.

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
```diff
@@ -0,0 +1,226 @@
+[EventPipe](/dotnet/core/diagnostics/eventpipe) is a runtime component used to collect tracing data, similar to [ETW](/windows/win32/etw/event-tracing-portal) and [perf_events](https://wikipedia.org/wiki/Perf_%28Linux%29).
+
+* Manual testing
+  * Browser developer tools: Download the `.json` output file, open the file in Visual Studio, and find the expected method calls.
```
it's not `.json` it's `.nettrace`

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
````diff
@@ -0,0 +1,226 @@
+</PropertyGroup>
+```
+
+Add Blazor start configuration in `wwwroot/index.html`, using the [fingerprinted location of the Blazor WebAssembly script](xref:blazor/fundamentals/static-files#fingerprint-client-side-static-assets-in-standalone-blazor-webassembly-apps). In the following example, the `sampleIntervalMs` option is set to 10 seconds, which is the default setting if `sampleIntervalMs` isn't specified:
````
You don't have to do `builder.withConfig` anymore, the MSbuild properties would suffice.

guardrex replied:
Are you asking to remove the following `Blazor.start` configuration? If so, how is the sample interval set?

guardrex replied:
Nevermind ... I think you just mean that this text should change to make this optional if they want to set the interval. I'll place this into a new section.

pavelsavara replied:
I mean that you could remove it and it should flow from MSbuild properties

guardrex replied:
> interval should flow from MSbuild properties

What's the property name/configuration? Is it .........

```xml
<SampleIntervalMs>10</SampleIntervalMs>
```

guardrex replied:
... and will the same thing be true for `callSpec`? Instead of `Blazor.start`, will the following work? ...

```xml
<CallSpec>N:BlazorSample</CallSpec>
```

guardrex replied:
so MSBuild for sample interval will be ....

```xml
<WasmProfilers>browser:sampleIntervalMs=10;</WasmProfilers>
```

guardrex replied:
Latest commit sets up guidance for both scenarios and both options. 🎉

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
```diff
@@ -0,0 +1,226 @@
+  * Browser developer tools: Download the `.json` output file, open the file in Visual Studio, and find the expected method calls.
+  * [`dotnet-trace`](/dotnet/core/diagnostics/dotnet-trace): Open the `.nettrace` output file in Visual Studio and find the expected method calls.
+* Web-based testing
+  * Use the JavaScript API to obtain a [promise](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise) of NetTrace (`.nettrace`) bytes.
```
Promise here is technical detail that most users would not need to know. Please drop mention of it.

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
````diff
@@ -0,0 +1,226 @@
+
+```xml
+<PropertyGroup>
+  <WasmProfilers>browser</WasmProfilers>
````
`<WasmProfilers>browser</WasmProfilers>` is not compatible with `<WasmPerfTracing>true</WasmPerfTracing>` at all. They are separate approach to profiling. When talking about ETW/EventPipe/DiagnostiServer the `WasmPerfTracing` is way to go.

pavelsavara commented on aspnetcore/blazor/performance/profiling.md:
```diff
@@ -0,0 +1,226 @@
+<PropertyGroup>
+  <WasmProfilers>browser</WasmProfilers>
+  <WasmPerfTracing>true</WasmPerfTracing>
+  <WasmPerfInstrumentation>true</WasmPerfInstrumentation>
```
`<WasmPerfInstrumentation>true</WasmPerfInstrumentation>` is making the app slower. People should yse it only when they want to do per profiling. This is not necessary for memory dump or counters.

pavelsavara reviewed (commented):
> I'm not on nightly builds. The untested examples are taken from your issue/PR remarks. I can go nightly if needed for testing, but I think you'll be able to suggest corrections based on what's on the PR.

I did the first pass. My main messages are
- please try it yourself.
- the dev tools profiling and EP/DS profiling are quite different set of tools and not compatible with each other. I would prefer them on separate page.
- log profiler should not be mentioned.

guardrex commented:
**UPDATE (4/25)**: I've been bogged down in security work all week. I should be able to get back to this early next week.

guardrex commented:
@pavelsavara ... Made the update pass. A few notes ......

* Not sure that I have the split right, noting again that I have very little experience with profiling.
* Not sure if "Event Pipe profiling" was the right way to focus one of the articles. You said "ETW/EventPipe/diagnostic server," so do we need to change the title, description, UID, and filename of the "Event Pipe" article to something more general?
* WRT running profiling locally, I don't know when I'll have time to do it. I'm buried in work these days. ⛰️⛏️
* I renamed your suggestion for the condition from `MyAppWithEnabledProfiler` to `ProfilingEnabled` to shorten it and also to avoid the use of "my" API language. I've avoided "my"-named things in API coverage for several years. You mentioned passing `/p:MyAppWithEnabledProfiler=true` (recasted on the PR to `/p:ProfilingEnabled=true`). I added that in Visual Studio the dev can add `<ProfilingEnabled>true</ProfilingEnabled>` to a VS publish profile ... is that correct? ... or did I just make that up 🙈😆? If I did just make it up, how can we tell VS users to disable the profiling with that condition on a publish out of VS?

pavelsavara commented:
>     * Not sure if "Event Pipe profiling" was the right way 

Maybe "Diagnostics with Event Pipe on WASM" or something like that.

>     * WRT running profiling locally

At least watch Daniel's demo of it please.

>     avoid the use of "my" API language to `/p:ProfilingEnabled=true`

The idea I'm trying to convey is that you _the user_ own that symbol (and it's conflicts with other stuff).

They could also name it `/p:WeWillPayTheExtraCpuCostWhenThisProfilingIsEnabled=true` 

> I added that in Visual Studio the dev can add `<ProfilingEnabled>true</ProfilingEnabled>`

Let's not promise things, I am not going to do it myself. Especially not with this specific symbol name.

pavelsavara commented on aspnetcore/blazor/performance/app-download-size.md:
````diff
@@ -89,7 +95,7 @@ A data file is included to make timezone information correct. If the app doesn't
 ```xml
 <PropertyGroup>
-<BlazorEnableTimeZoneSupport>false</BlazorEnableTimeZoneSupport>
+  <BlazorEnableTimeZoneSupport>false</BlazorEnableTimeZoneSupport>
````
we have `InvariantTimezone` and `BlazorEnableTimeZoneSupport`

@maraf do we have preference. Do both work ?

guardrex replied:
I found out that we're supposed to only be describing `<InvariantTimezone>`, per guidance in our Glob-Loc article ...

> > [!NOTE]
> `<BlazorEnableTimeZoneSupport>` overrides an earlier `<InvariantTimezone>` setting. We recommend removing the `<BlazorEnableTimeZoneSupport>` setting.

pavelsavara commented on aspnetcore/blazor/performance/developer-tools-profiling.md:
```diff
@@ -1,22 +1,22 @@
+title: ASP.NET Core Blazor WebAssembly developer tools performance profiling and diagnostic counters
 author: guardrex
-description: Learn about performance profiling and diagnostic counters in ASP.NET Core Blazor WebAssembly apps.
+description: Learn about developer tools performance profiling and diagnostic counters in ASP.NET Core Blazor WebAssembly apps.
```
It's "browser developer tools" a normally hidden panel in the browser.
It doesn't support "diagnostic counters" only "sampling profiler" via code instrumentation.

pavelsavara commented on aspnetcore/blazor/performance/developer-tools-profiling.md:
```diff
@@ -48,8 +47,12 @@ Property | Default | Set value to&hellip; | Description
 `<WasmNativeDebugSymbols>` | `true` | `true` | Controls building with native debug symbols.
 `<WasmBuildNative>` | `false` | `true` | Controls building the native executable.
 
+Enabling profilers has negative size and performance impact, so don't publish an app for production with profilers enabled. In the following example, a condition is set on a property group section that only enables profiling when the app is built with `/p:ProfilingEnabled=true` (.NET CLI) or `<ProfilingEnabled>true</ProfilingEnabled>` in a Visual Studio publish profile.
```
The other alternative is to enumerate all the features on commandline, when you want have it anabled.

Like `/p:WasmProfilers=browser`

pavelsavara commented on aspnetcore/blazor/performance/event-pipe-profiling.md:
```diff
@@ -0,0 +1,119 @@
+ms.date: 04/29/2025
+uid: blazor/performance/event-pipe-profiling
+---
+# ASP.NET Core Blazor WebAssembly Event Pipe performance profiling and diagnostic counters
```
"Event Pipe" can also give you GC heap dump. It's a major feature.
But I'm not sure we need all  features in the title.

Perhaps simpler "ASP.NET Core Blazor WebAssembly diagnostic with Event Pipe"

pavelsavara commented on aspnetcore/blazor/performance/browser-developer-tools-profiling.md:
```diff
@@ -1,22 +1,22 @@
 ---
-title: ASP.NET Core Blazor WebAssembly developer tools performance profiling and diagnostic counters
+title: ASP.NET Core Blazor WebAssembly browser developer tools diagnostic profiling
```
please drop `profiling`. I would not call GC dump or counters "profiling". But all of it is diagnostics in my mind.

pavelsavara commented on aspnetcore/blazor/performance/browser-developer-tools-profiling.md:
```diff
@@ -47,21 +47,25 @@ Property | Default | Set value to&hellip; | Description
-  <RunAOTCompilationAfterBuild>true</RunAOTCompilationAfterBuild>
   <WasmNativeStrip>false</WasmNativeStrip>
   <WasmNativeDebugSymbols>true</WasmNativeDebugSymbols>
   <WasmBuildNative>true</WasmBuildNative>
```
WasmBuildNative is not required either

pavelsavara commented on aspnetcore/blazor/performance/browser-developer-tools-profiling.md:
````diff
@@ -47,21 +47,25 @@ Property | Default | Set value to&hellip; | Description
+Alternatively, enable features when the app is built with the .NET CLI. The following options passed to the `dotnet build` command mirror the preceding MS Build property configuration:
+
+```dotnetcli
+/p:WasmProfilers=browser /p:WasmNativeStrip=false /p:WasmNativeDebugSymbols=true /p:WasmBuildNative=true
````
please drop WasmBuildNative

guardrex commented:
@pavelsavara ...

Articles are now ...

Browser developer tools:

* Title: ASP.NET Core Blazor WebAssembly browser developer tools diagnostics
* Description: Learn about browser developer tools diagnostics in ASP.NET Core Blazor WebAssembly apps.
* UID: `blazor/performance/webassembly-browser-developer-tools`
* Filename/URL path: `performance/webassembly-browser-developer-tools-diagnostics.md`

Event Pipe:

* Title: ASP.NET Core Blazor WebAssembly Event Pipe diagnostics
* Description: Learn about Event Pipe diagnostics and how to get a Garbage Collector heap dump in ASP.NET Core Blazor WebAssembly apps.
* UID: `blazor/performance/webassembly-event-pipe`
* Filename/URL path: `performance/webassembly-event-pipe-diagnostics.md`

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,202 @@
+
+App code can be manually profiled using the diagnostic profiler in a browser's developer tools console.
+
+Built-in diagnostic counters are available to track:
```
counters -> features

guardrex replied:
```suggestion
Built-in diagnostic features are available to track:
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,202 @@
+Built-in diagnostic counters are available to track:
+
+* [Ahead-of-time (AOT) compilation](xref:blazor/tooling/webassembly#ahead-of-time-aot-compilation)
+* Code interpolation
```
"Code interpolation" is not a thing

guardrex replied:
```suggestion
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,202 @@
+Property | Default | Set value to&hellip; | Description
+--- | :---: | :---: | ---
+`<WasmProfilers>` | No value | `browser` | Mono profilers to use. Potential values are "`browser`" and "`log`". To use both, separate the values with a semicolon. The `browser` profiler enables integration with the browser's developer tools profiler.
+`<RunAOTCompilation>`| `false` | `true` | Controls AOT compilation.
```
drop RunAOTCompilation, RunAOTCompilationAfterBuild, WasmBuildNative here

guardrex commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
```diff
@@ -0,0 +1,125 @@
+Built-in diagnostic counters are available to track:
+
+* [Ahead-of-time (AOT) compilation](xref:blazor/tooling/webassembly#ahead-of-time-aot-compilation)
+* Code interpolation
```
```suggestion
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
````diff
@@ -0,0 +1,202 @@
+/p:WasmProfilers=browser /p:WasmNativeStrip=false /p:WasmNativeDebugSymbols=true
+```
+
+Setting WebAssembly profilers with `<WasmProfilers>browser;</WasmProfilers>` doesn't require AOT (`<RunAOTCompilation>`/`<RunAOTCompilationAfterBuild>` set to `false` or removed from the preceding property group).
````
"doesn't require AOT" is good, I would stop there.

guardrex replied:
```suggestion
Setting WebAssembly profilers with `<WasmProfilers>browser;</WasmProfilers>` doesn't require AOT.
```

guardrex commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,202 @@
+Property | Default | Set value to&hellip; | Description
+--- | :---: | :---: | ---
+`<WasmProfilers>` | No value | `browser` | Mono profilers to use. Potential values are "`browser`" and "`log`". To use both, separate the values with a semicolon. The `browser` profiler enables integration with the browser's developer tools profiler.
+`<RunAOTCompilation>`| `false` | `true` | Controls AOT compilation.
+`<RunAOTCompilationAfterBuild>` | `false` | `true` | Controls AOT compilation after build. By default, it's run only for publish.
```
```suggestion
```

guardrex commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,202 @@
+`<RunAOTCompilationAfterBuild>` | `false` | `true` | Controls AOT compilation after build. By default, it's run only for publish.
+`<WasmNativeStrip>` | `true` | `false` | Controls stripping the native executable.
+`<WasmNativeDebugSymbols>` | `true` | `true` | Controls building with native debug symbols.
+`<WasmBuildNative>` | `false` | `true` | Controls building the native executable.
```
```suggestion
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
````diff
@@ -0,0 +1,202 @@
+</PropertyGroup>
+```
+
+Alternatively, add the following Blazor start configuration in `wwwroot/index.html` and add `autostart="false"` to the Blazor `<script>` tag. The `{INTERVAL}` placeholder represents the time in milliseconds:
````
Please drop the whole paragraph about Blazor.start

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,202 @@
+
+The following example:
+
+* Collects a GC (Garbage Collector) dump of the live .NET process.
```
and again

guardrex replied:
```suggestion
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,202 @@
+`<WasmProfilers>` | No value | `browser` | Mono profilers to use. Potential values are "`browser`" and "`log`". To use both, separate the values with a semicolon. The `browser` profiler enables integration with the browser's developer tools profiler.
+`<WasmPerfTracing>` | `false` | `true` | Controls diagnostic server tracing.
+`<WasmPerfInstrumentation>` | `false` | `true` | Controls CPU sampling instrumentation for diagnostic server.
+`<MetricsSupport>` | `false` | `true` | Controls `System.Diagnostics.Metrics` support. For more information, see the [`System.Diagnostics.Metrics` namespace](/dotnet/api/system.diagnostics.metrics).
```
Drop MetricsSupport and EventSourceSupport, they are unrelated to this page. They make sense in EP page.

guardrex commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
````diff
@@ -0,0 +1,202 @@
+</PropertyGroup>
+```
+
+Alternatively, add the following Blazor start configuration in `wwwroot/index.html` and add `autostart="false"` to the Blazor `<script>` tag. The `{INTERVAL}` placeholder represents the time in milliseconds:
+
+```html
+<script src="_framework/blazor.webassembly#[.{fingerprint}].js" 
+    autostart="false"></script>
+<script>
+  Blazor.start({
+    configureRuntime: function (builder) {
+      builder.withConfig({
+        browserProfilerOptions: {
+          sampleIntervalMs: {INTERVAL},
+        }
+      });
+    }
+  });
+</script>
+```
+
````
```suggestion
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,202 @@
+
+The [`Timing-Allow-Origin` HTTP header](https://developer.mozilla.org/docs/Web/HTTP/Reference/Headers/Timing-Allow-Origin) allows for more precise time measurements.
+
+Browser developer tools console calls in the following example that trigger profiling:
```
Sorry, this must be confusing. Those commands do not belong to this page, but to EP page.

globalThis.getDotnetRuntime(0).collectGcDump();
globalThis.getDotnetRuntime(0).collectPerfCounters({durationSeconds: 60});
globalThis.getDotnetRuntime(0).collectCpuSamples({durationSeconds: 60});

guardrex commented:
@pavelsavara ... That round of updates is applied.

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
````diff
@@ -0,0 +1,145 @@
+```xml
+<PropertyGroup>
+  <WasmProfilers>browser</WasmProfilers>
+  <WasmPerfTracing>true</WasmPerfTracing>
````
WasmPerfTracing doesn't belong on this page

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,145 @@
+Property | Default | Set value to&hellip; | Description
+--- | :---: | :---: | ---
+`<WasmProfilers>` | No value | `browser` | Mono profilers to use. Potential values are "`browser`" and "`log`". To use both, separate the values with a semicolon. The `browser` profiler enables integration with the browser's developer tools profiler.
+`<WasmPerfTracing>` | `false` | `true` | Controls diagnostic server tracing.
```
drop WasmPerfTracing

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
```diff
@@ -0,0 +1,122 @@
+* Manual testing
+  * Browser developer tools: Download the `.nettrace` output file, open the file in Visual Studio, and find the expected method calls.
+  * [`dotnet-trace`](/dotnet/core/diagnostics/dotnet-trace): Open the `.nettrace` output file in Visual Studio and find the expected method calls.
+* Web-based testing
```
drop "Web-based testing' paragraph, this was probably notes about unit testing for our team

pavelsavara replied:
still here

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
```diff
@@ -0,0 +1,122 @@
+  * Upload the file via HTTP.
+  * Parse and validate that the trace contains the expected method calls.
+
+Built-in diagnostic counters are available to track [Ahead-of-time (AOT) compilation](xref:blazor/tooling/webassembly#ahead-of-time-aot-compilation).
```
drop AOT mention here

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
```diff
@@ -0,0 +1,122 @@
+* Manual testing
+  * Browser developer tools: Download the `.json` output file, open the file in Visual Studio, and find the expected classes.
+  * [`dotnet-gcdump` (`collect`/convert` options)](/dotnet/core/diagnostics/dotnet-gcdump): To view the captured GC dump files, see [View the GC dump captured from dotnet-gcdump](/dotnet/core/diagnostics/dotnet-gcdump#view-the-gc-dump-captured-from-dotnet-gcdump).
+* Web-based testing
```
drop

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
```diff
@@ -0,0 +1,122 @@
+* Manual testing
+  * Browser developer tools: Download the `.json` output file, open the file in Visual Studio, and find the expected counters.
+  * [`dotnet-counters collect`](/dotnet/core/diagnostics/dotnet-counters): Open the `.csv`/`.json` output file in Visual Studio and find the expected counters.
+* Web-based testing
```
drop

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
````diff
@@ -0,0 +1,122 @@
+* `collectCpuSamples(durationSeconds)`: Collect CPU diagnostic counters.
+
+```javascript
+globalThis.getDotnetRuntime(0).collectGcDump();
````
each of these commands is the way how you get the .nettrace, so they could be in each of paragraphs above

guardrex commented:
@pavelsavara ... Latest round of updates complete.

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,125 @@
+
+The following example:
+
+* Collects performance counters for 60 seconds.
```
drop counters, Collects CPU counters

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,125 @@
+
+Property | Default | Set value to&hellip; | Description
+--- | :---: | :---: | ---
+`<WasmProfilers>` | No value | `browser` | Mono profilers to use. Potential values are "`browser`" and "`log`". To use both, separate the values with a semicolon. The `browser` profiler enables integration with the browser's developer tools profiler.
```
drop `log`

guardrex replied:
```suggestion
`<WasmProfilers>` | No value | `browser` | Mono profiler. Currently, only "`browser`" is supported. The `browser` profiler enables integration with the browser's developer tools profiler.
```

guardrex commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,125 @@
+
+[!INCLUDE[](~/includes/aspnetcore-repo-ref-source-links.md)]
+
+The following example:
+
+* Collects performance counters for 60 seconds.
+* Collects CPU counters for 60 seconds.
+
```
```suggestion
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
```diff
@@ -0,0 +1,124 @@
+
+[EventPipe](/dotnet/core/diagnostics/eventpipe) is a runtime component used to collect tracing data, similar to [ETW](/windows/win32/etw/event-tracing-portal) and [perf_events](https://wikipedia.org/wiki/Perf_%28Linux%29).
+
+* Manual testing
```
this is notes for my team

guardrex replied:
Not sure what you're asking me to do here ... delete? ... and if so, delete what (which lines)?

pavelsavara replied:
delete whole Manual testing and Web-based testing

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
````diff
@@ -0,0 +1,124 @@
+
+* `collectPerfCounters(durationSeconds)`: Collect general diagnostic counters.
+
+    ```javascript
````
you have dedicated paragraph above, why you don't move it ?

pavelsavara commented:
@guardrex anyone at home ? am I talking to copilot today ?

guardrex commented:
I didn't get an email ping on your last comment.

The only thing left that I'm aware of is ...

> this is notes for my team

I'm not sure which lines to delete.

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
````diff
@@ -0,0 +1,116 @@
+Collect a GC (Garbage Collector) dump of the live .NET process with `collectGcDump`:
+
+```javascript
+globalThis.getDotnetRuntime(0).collectGcDump();
````
move this to gc dump paragraph

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
````diff
@@ -0,0 +1,116 @@
+Collect diagnostic counters for 60 seconds with `collectPerfCounters(durationSeconds)`:
+
+```javascript
+globalThis.getDotnetRuntime(0).collectPerfCounters({durationSeconds: 60});
````
nove this to counters paragraph

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
````diff
@@ -0,0 +1,116 @@
+Collect CPU counters for 60 seconds with `collectCpuSamples(durationSeconds)`:
+
+```javascript
+globalThis.getDotnetRuntime(0).collectCpuSamples({durationSeconds: 60});
````
move this to profiling paragraph

guardrex commented:
***Updates made!***

This was one of the very rare cases where my attempt to proactively resolve a preview issue to save time and effort failed. 😢 It shouldn't be a problem going forward under the new plan for preview feature documentation.

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
````diff
@@ -0,0 +1,109 @@
+globalThis.getDotnetRuntime(0).collectPerfCounters({durationSeconds: 60});
+```
+
+## .NET Core Diagnostics Client Library example
````
Maybe move the whole paragraph with the msbuild properties up in the page, just after prerequisites, because you can;t really use the `globalThis.getDotnetRuntime(0).collectPerfCounters` before enabling that.
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
```diff
@@ -0,0 +1,109 @@
+
+## GC (Garbage Collector) dumps
+
+Manual testing:
```
Manual testing

pavelsavara commented on aspnetcore/blazor/performance/webassembly-event-pipe-diagnostics.md:
```diff
@@ -0,0 +1,109 @@
+
+Manual testing:
+
+* Browser developer tools: Download the `.json` output file, open the file in Visual Studio, and find the expected classes.
```
json

pavelsavara replied:
all locations please

guardrex commented:
***Updates made!*** ... and I make some improvements to the layout and content of those sections.

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
````diff
@@ -0,0 +1,120 @@
+
+```xml
+<PropertyGroup>
+  <WasmProfilers>browser:sampleIntervalMs={INTERVAL};</WasmProfilers>
````
it's `interval` not `sampleIntervalMs`

guardrex replied:
```suggestion
  <WasmProfilers>browser:interval={INTERVAL};</WasmProfilers>
```

pavelsavara commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
```diff
@@ -0,0 +1,120 @@
+
+## .NET Core Diagnostics Client Library example
+
+Parse and validate NetTrace (`.nettrace`) messages using the .NET Core Diagnostics Client Library:
```
`Parse and validate NetTrace` is EP related, doesn't belon to this page. Drop lines up to 100

guardrex commented on aspnetcore/blazor/performance/webassembly-browser-developer-tools-diagnostics.md:
````diff
@@ -0,0 +1,120 @@
+```xml
+<PropertyGroup>
+  <WasmProfilers>browser</WasmProfilers>
+  <WasmPerfInstrumentation>true</WasmPerfInstrumentation>
````
```suggestion
```

guardrex commented:
@pavelsavara ... Updates made.

pavelsavara commented:
let's merge this
